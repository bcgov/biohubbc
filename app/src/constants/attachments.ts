export enum AttachmentType {
  REPORT = 'Report',
  OTHER = 'Other'
}

export enum ProjectAttachmentValidExtensions {
  AUDIO = '.wav, .mp3, .mp4, .wma',
  DATA = '.txt, .xls, .xlsx, .xlsm, .xlsb, .accdb, .mdb, .ods, .csv',
  IMAGE = '.gif, .png, .jpg, .jpeg, .svg, .tiff, .bmp, .tif',
  REPORT = '.doc, .docx, .pdf',
  SPATIAL = '.kml, .gpx, .zip',
  VIDEO = '.mp4, .mov, .wmv, .ave',
  OTHER = '.wav, .mp3, .mp4, .wma, .txt, .xls, .xlsx, .xlsm, .xlsb, .accdb, .mdb, .ods, .csv, .gif, .png, .jpg, .jpeg, .svg, .tiff, .bmp, .tif, .doc, .docx, .kml, .gdb, .lyr, .shp, .las, .laz, .zlas, .gml, .gpx, .json, .mp4, .mov, .wmv, .ave'
}
