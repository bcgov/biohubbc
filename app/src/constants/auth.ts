export enum SYSTEM_IDENTITY_SOURCE {
  BCEID_BUSINESS = 'BCEIDBUSINESS',
  BCEID_BASIC = 'BCEIDBASIC',
  IDIR = 'IDIR',
  DATABASE = 'DATABASE',
  UNVERIFIED = 'UNVERIFIED'
}
