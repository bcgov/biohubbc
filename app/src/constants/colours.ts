import { Color } from '@mui/material';
import {
  blue,
  blueGrey,
  brown,
  cyan,
  deepOrange,
  deepPurple,
  green,
  grey,
  indigo,
  orange,
  pink,
  purple,
  red,
  teal
} from '@mui/material/colors';

type ColourMap = Record<string, { colour: Color }>;

/**
 * Default colour fallback.
 *
 */
const DEFAULT_COLOUR = blueGrey;

/**
 * Colour map for `Survey Progress` chips.
 *
 */
const SURVEY_PROGRESS_COLOUR_MAP = {
  Planning: { colour: blueGrey },
  'In progress': { colour: deepPurple },
  Completed: { colour: green }
};

/**
 * Colour map for `Taxon Rank` chips.
 *
 */
const TAXON_RANK_COLOUR_MAP = {
  Subspecies: { colour: blue },
  Variety: { colour: blue },
  Species: { colour: purple },
  Genus: { colour: teal },
  Family: { colour: red },
  Order: { colour: indigo },
  Class: { colour: deepOrange },
  Phylum: { colour: pink },
  Kingdom: { colour: grey }
};

/**
 * Colour map for `NRM Region` chips.
 *
 */
const NRM_REGION_COLOUR_MAP = {
  'Kootenay-Boundary Natural Resource Region': { colour: cyan },
  'Thompson-Okanagan Natural Resource Region': { colour: orange },
  'West Coast Natural Resource Region': { colour: green },
  'Cariboo Natural Resource Region': { colour: deepPurple },
  'South Coast Natural Resource Region': { colour: blue },
  'Northeast Natural Resource Region': { colour: brown },
  'Omineca Natural Resource Region': { colour: pink },
  'Skeena Natural Resource Region': { colour: red }
};

/**
 * Generate colour getter from ColourMap.
 *
 * @param {ColourMap} colourMap
 * @param {Color} [fallbackColour]
 * @returns {(lookup: string) => Color} Colour getter
 */
const generateColourMapGetter = <T extends ColourMap>(colourMap: T, fallbackColour = DEFAULT_COLOUR) => {
  return (lookup: string) => colourMap[lookup]?.colour ?? fallbackColour;
};

/**
 * Get survey progress colour mapping.
 *
 */
export const getSurveyProgressColour = generateColourMapGetter(SURVEY_PROGRESS_COLOUR_MAP);

/**
 * Get taxon rank colour mapping.
 *
 */
export const getTaxonRankColour = generateColourMapGetter(TAXON_RANK_COLOUR_MAP);

/**
 * Get NRM region colour mapping.
 *
 */
export const getNrmRegionColour = generateColourMapGetter(NRM_REGION_COLOUR_MAP);
