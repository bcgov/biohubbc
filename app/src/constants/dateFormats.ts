// See BC Gov standards: https://www2.gov.bc.ca/gov/content/governments/services-for-government/policies-procedures/web-content-development-guides/writing-for-the-web/web-style-guide/numbers

export const ShortDateFormat = 'YYYY-MM-DD'; //2020-01-05

export const ShortDateTimeFormat = 'YYYY-MM-DD, H:mm a'; //2020-01-05, 3:30 pm

export const MediumDateFormat = 'MMMM D, YYYY'; //January 5, 2020

export const MediumDateTimeFormat = 'MMMM D, YYYY, H:mm a'; //January 5, 2020, 3:30 pm

export const LongDateFormat = 'dddd, MMMM D, YYYY, H:mm a'; //Monday, January 5, 2020, 3:30 pm

export const LongDateTimeFormat = 'dddd, MMMM D, YYYY, H:mm a'; //Monday, January 5, 2020, 3:30 pm
