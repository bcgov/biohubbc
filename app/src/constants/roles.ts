export const Roles = {
  SYSTEM_ADMINISTRATOR: 'ADM'
};
