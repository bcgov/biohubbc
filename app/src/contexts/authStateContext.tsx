import useKeycloakWrapper from 'hooks/useKeycloakWrapper';
import React from 'react';

export interface IAuthState {
  ready?: boolean;
}

export const AuthStateContext = React.createContext<IAuthState>({
  ready: false
});

export const AuthStateContextProvider: React.FC = (props) => {
  const keycloakWrapper = useKeycloakWrapper();

  return (
    <AuthStateContext.Provider value={{ ready: keycloakWrapper.keycloak?.authenticated }}>
      {props.children}
    </AuthStateContext.Provider>
  );
};
