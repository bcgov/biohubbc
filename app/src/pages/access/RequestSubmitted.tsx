import { mdiCheck } from '@mdi/js';
import Icon from '@mdi/react';
import Box from '@mui/material/Box';
import Button from '@mui/material/Button';
import CircularProgress from '@mui/material/CircularProgress';
import Container from '@mui/material/Container';
import Typography from '@mui/material/Typography';
import { useAuthStateContext } from 'hooks/useAuthStateContext';
import { Redirect } from 'react-router';

const RequestSubmitted = () => {
  const authStateContext = useAuthStateContext();

  if (authStateContext.simsUserWrapper.isLoading) {
    // User data has not been loaded, can not yet determine if they have a role
    return <CircularProgress className="pageProgress" />;
  }

  if (!authStateContext.simsUserWrapper.hasAccessRequest || authStateContext.simsUserWrapper.roleNames?.length) {
    // User has no pending access request or already has a role
    return <Redirect to={{ pathname: '/' }} />;
  }

  return (
    <Container>
      <Box pt={6} textAlign="center">
        <Icon path={mdiCheck} size={2} color="#4caf50" />
        <h1>Access Request Submitted</h1>
        <Typography>Your request is currently pending a review by an administrator.</Typography>
        <Box pt={4}>
          <Button
            component="a"
            onClick={() => authStateContext.auth.signoutRedirect()}
            type="submit"
            size="large"
            variant="contained"
            color="primary"
            data-testid="logout-button">
            Log Out
          </Button>
        </Box>
      </Box>
    </Container>
  );
};

export default RequestSubmitted;
