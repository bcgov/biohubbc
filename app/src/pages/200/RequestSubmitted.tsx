import Box from '@material-ui/core/Box';
import Button from '@material-ui/core/Button';
import CircularProgress from '@material-ui/core/CircularProgress';
import Container from '@material-ui/core/Container';
import Typography from '@material-ui/core/Typography';
import { mdiCheck } from '@mdi/js';
import Icon from '@mdi/react';
import { AuthStateContext } from 'contexts/authStateContext';
import React, { useContext } from 'react';
import { Redirect, useHistory } from 'react-router';

const RequestSubmitted = () => {
  const history = useHistory();

  const { keycloakWrapper } = useContext(AuthStateContext);

  if (!keycloakWrapper?.hasLoadedAllUserInfo) {
    // User data has not been loaded, can not yet determine if they have a role
    return <CircularProgress className="pageProgress" />;
  }

  if (keycloakWrapper?.systemRoles.length) {
    // User already has a role
    return <Redirect to={{ pathname: '/admin/projects' }} />;
  }

  if (!keycloakWrapper.hasAccessRequest) {
    // User has no pending access request
    return <Redirect to={{ pathname: '/' }} />;
  }

  return (
    <Container>
      <Box pt={6} textAlign="center">
        <Icon path={mdiCheck} size={2} color="#4caf50" />
        <h1>Access Request Submitted</h1>
        <Typography>Your access request has been submitted for review.</Typography>
        <Box pt={4}>
          <Button
            onClick={() => {
              history.push('/logout');
            }}
            type="submit"
            size="large"
            variant="contained"
            color="primary"
            data-testid="logout-button">
            Log Out
          </Button>
        </Box>
      </Box>
    </Container>
  );
};

export default RequestSubmitted;
