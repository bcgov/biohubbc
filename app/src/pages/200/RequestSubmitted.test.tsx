import { fireEvent, render, waitFor } from '@testing-library/react';
import { SYSTEM_ROLE } from 'constants/roles';
import { AuthStateContext } from 'contexts/authStateContext';
import { createMemoryHistory } from 'history';
import React from 'react';
import { Router } from 'react-router';
import RequestSubmitted from './RequestSubmitted';

describe('RequestSubmitted', () => {
  it('renders a spinner when `hasLoadedAllUserInfo` is false', () => {
    const authState = {
      keycloakWrapper: {
        hasLoadedAllUserInfo: false,
        systemRoles: [],
        hasAccessRequest: false,

        keycloak: {},
        getUserIdentifier: jest.fn(),
        hasSystemRole: jest.fn(),
        getIdentitySource: jest.fn(),
        username: 'testusername',
        displayName: 'testdisplayname',
        email: 'test@email.com',
        firstName: 'testfirst',
        lastName: 'testlast',
        refresh: () => {}
      }
    };

    const history = createMemoryHistory();

    history.push('/access-request');

    const { asFragment } = render(
      <AuthStateContext.Provider value={authState}>
        <Router history={history}>
          <RequestSubmitted />
        </Router>
      </AuthStateContext.Provider>
    );

    // does not change location
    expect(history.location.pathname).toEqual('/access-request');

    // renders a spinner
    expect(asFragment()).toMatchSnapshot();
  });

  it('redirects to `/admin/projects` when user has at least 1 system role', () => {
    const authState = {
      keycloakWrapper: {
        hasLoadedAllUserInfo: true,
        systemRoles: [SYSTEM_ROLE.PROJECT_CREATOR],
        hasAccessRequest: false,

        keycloak: {},
        getUserIdentifier: jest.fn(),
        hasSystemRole: jest.fn(),
        getIdentitySource: jest.fn(),
        username: 'testusername',
        displayName: 'testdisplayname',
        email: 'test@email.com',
        firstName: 'testfirst',
        lastName: 'testlast',
        refresh: () => {}
      }
    };

    const history = createMemoryHistory();

    history.push('/access-request');

    render(
      <AuthStateContext.Provider value={authState}>
        <Router history={history}>
          <RequestSubmitted />
        </Router>
      </AuthStateContext.Provider>
    );

    expect(history.location.pathname).toEqual('/admin/projects');
  });

  it('redirects to `/` when user has no pending access request', () => {
    const authState = {
      keycloakWrapper: {
        hasLoadedAllUserInfo: true,
        systemRoles: [],
        hasAccessRequest: false,

        keycloak: {},
        getUserIdentifier: jest.fn(),
        hasSystemRole: jest.fn(),
        getIdentitySource: jest.fn(),
        username: 'testusername',
        displayName: 'testdisplayname',
        email: 'test@email.com',
        firstName: 'testfirst',
        lastName: 'testlast',
        refresh: () => {}
      }
    };

    const history = createMemoryHistory();

    history.push('/access-request');

    render(
      <AuthStateContext.Provider value={authState}>
        <Router history={history}>
          <RequestSubmitted />
        </Router>
      </AuthStateContext.Provider>
    );

    expect(history.location.pathname).toEqual('/');
  });

  it('renders correctly when user has no role but has a pending access requests', () => {
    const authState = {
      keycloakWrapper: {
        hasLoadedAllUserInfo: true,
        systemRoles: [],
        hasAccessRequest: true,

        keycloak: {},
        getUserIdentifier: jest.fn(),
        hasSystemRole: jest.fn(),
        getIdentitySource: jest.fn(),
        username: 'testusername',
        displayName: 'testdisplayname',
        email: 'test@email.com',
        firstName: 'testfirst',
        lastName: 'testlast',
        refresh: () => {}
      }
    };

    const history = createMemoryHistory();

    history.push('/access-request');

    const { getByText, asFragment } = render(
      <AuthStateContext.Provider value={authState}>
        <Router history={history}>
          <RequestSubmitted />
        </Router>
      </AuthStateContext.Provider>
    );

    // does not change location
    expect(history.location.pathname).toEqual('/access-request');

    expect(getByText('Log Out')).toBeVisible();

    // renders the component in full
    expect(asFragment()).toMatchSnapshot();
  });

  describe('Log Out', () => {
    const history = createMemoryHistory();

    it('should redirect to `/logout`', async () => {
      const authState = {
        keycloakWrapper: {
          hasLoadedAllUserInfo: true,
          systemRoles: [],
          hasAccessRequest: true,

          keycloak: {},
          getUserIdentifier: jest.fn(),
          hasSystemRole: jest.fn(),
          getIdentitySource: jest.fn(),
          username: 'testusername',
          displayName: 'testdisplayname',
          email: 'test@email.com',
          firstName: 'testfirst',
          lastName: 'testlast',
          refresh: () => {}
        }
      };

      const { getByTestId } = render(
        <AuthStateContext.Provider value={authState}>
          <Router history={history}>
            <RequestSubmitted />
          </Router>
        </AuthStateContext.Provider>
      );

      fireEvent.click(getByTestId('logout-button'));

      waitFor(() => {
        expect(history.location.pathname).toEqual('/logout');
      });
    });
  });
});
