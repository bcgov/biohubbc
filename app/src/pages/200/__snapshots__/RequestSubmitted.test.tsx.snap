// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RequestSubmitted renders a spinner when \`hasLoadedAllUserInfo\` is false 1`] = `
<DocumentFragment>
  <div
    class="MuiCircularProgress-root pageProgress MuiCircularProgress-colorPrimary MuiCircularProgress-indeterminate"
    role="progressbar"
    style="width: 40px; height: 40px;"
  >
    <svg
      class="MuiCircularProgress-svg"
      viewBox="22 22 44 44"
    >
      <circle
        class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
        cx="44"
        cy="44"
        fill="none"
        r="20.2"
        stroke-width="3.6"
      />
    </svg>
  </div>
</DocumentFragment>
`;

exports[`RequestSubmitted renders correctly when user has no role but has a pending access requests 1`] = `
<DocumentFragment>
  <div
    class="MuiContainer-root MuiContainer-maxWidthLg"
  >
    <div
      class="MuiBox-root MuiBox-root-1"
    >
      <svg
        role="presentation"
        style="width: 3rem; height: 3rem;"
        viewBox="0 0 24 24"
      >
        <path
          d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"
          style="fill: #4caf50;"
        />
      </svg>
      <h1>
        Access Request Submitted
      </h1>
      <p
        class="MuiTypography-root MuiTypography-body1"
      >
        Your access request has been submitted for review.
      </p>
      <div
        class="MuiBox-root MuiBox-root-2"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-containedSizeLarge MuiButton-sizeLarge"
          data-testid="logout-button"
          tabindex="0"
          type="submit"
        >
          <span
            class="MuiButton-label"
          >
            Log Out
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
